extensions/parquet/table/src/main/java/io/deephaven/parquet/table/pagestore/OffsetIndexBasedColumnChunkPageStore.java

```diff
@@ -0,0 +1,147 @@
+     * row to be coming from last page.
+     */
+    private final long fixedPageSize;
+
+    private final class PageState {
+        WeakReference<PageCache.IntrusivePage<ATTR>> pageRef;
+
+        PageState() {
+            // Initialized when used for the first time
+            pageRef = null;
+        }
+    }
+
+    private final AtomicReferenceArray<PageState> pageStates;
+    private final ColumnChunkReader.ColumnPageDirectAccessor columnPageDirectAccessor;
+
+    OffsetIndexBasedColumnChunkPageStore(@NotNull final PageCache<ATTR> pageCache,
+            @NotNull final ColumnChunkReader columnChunkReader,
+            final long mask,
+            @NotNull final ToPage<ATTR, ?> toPage) throws IOException {
+        super(pageCache, columnChunkReader, mask, toPage);
+        offsetIndex = columnChunkReader.getOffsetIndex();
+        Assert.assertion(offsetIndex != null, "offsetIndex != null");
+        numPages = offsetIndex.getPageCount();
+        Assert.assertion(numPages > 0, "numPages > 0");
+        pageStates = new AtomicReferenceArray<>(numPages);
+        columnPageDirectAccessor = columnChunkReader.getPageAccessor();
+
+        if (numPages == 1) {
+            fixedPageSize = numRows();
+            return;
+        }
+        boolean isPageSizeFixed = true;
+        final long firstPageSize = offsetIndex.getFirstRowIndex(1) - offsetIndex.getFirstRowIndex(0);
+        for (int i = 2; i < numPages && isPageSizeFixed; ++i) {
+            if (offsetIndex.getFirstRowIndex(i) - offsetIndex.getFirstRowIndex(i - 1) != firstPageSize) {
+                isPageSizeFixed = false;
+            }
+        }
+        fixedPageSize = isPageSizeFixed ? firstPageSize : PAGE_SIZE_NOT_FIXED;
+    }
+
+    /**
+     * Binary search in offset index to find the page number that contains the row. Logic duplicated from
+     * {@link Arrays#binarySearch(long[], long)} to use the offset index.
+     */
+    private static int findPageNumUsingOffsetIndex(final OffsetIndex offsetIndex, final long row) {
+        int low = 0;
+        int high = offsetIndex.getPageCount() - 1;
+
+        while (low <= high) {
+            final int mid = (low + high) >>> 1;
+            final long midVal = offsetIndex.getFirstRowIndex(mid);
+
+            if (midVal < row)
+                low = mid + 1;
+            else if (midVal > row)
+                high = mid - 1;
+            else
+                return mid; // key found
+        }
+        return -(low + 1); // key not found.
+    }
+
+    private ChunkPage<ATTR> getPage(final int pageNum) {
+        if (pageNum < 0 || pageNum >= numPages) {
+            throw new IllegalArgumentException("pageNum " + pageNum + " is out of range [0, " + numPages + ")");
+        }
+        PageCache.IntrusivePage<ATTR> page;
+        PageState pageState = pageStates.get(pageNum);
+        if (pageState == null) {
+            pageState = pageStates.updateAndGet(pageNum, p -> p == null ? new PageState() : p);
+        }
```

Comment:
While this pattern does work, it's a bit sloppy IMO; we should be able to implement the same sort of loop-CASing here a bit more efficiently.

Reply:
I'm not worried about the atomic array; I think that `get` + `updateAndGet` is a bit sloppy. We can write tighter code that accomplishes more specifically what we want. Look at the implementation for `updateAndGet` - happy to walk through this with you if you want.

Reply:
The other approach to do this would look like:
```
PageState pageState = pageStates.get(pageNum);
if (pageState == null && !pageStates.compareAndSet(pageNum, null, pageState = new PageState()) {
    pageState = pageStates.get(pageNum);
} 
```
Do you think this would be better than `updateAndGet`?

Reply:
Well, there will need to be a loop, this is not correct w/ CAS semantics. The issue is that as written, in the case where our inner function `p != null`, we're doing a spurious volatile write.

Something like:

```java
PageState pageState;
while ((pageState = pageStates.get(pageNum)) == null) {
   pageState = new PageState();
   if (pageStates.weakCompareAndSetVolatile(pageNum, null, pageState)) {
     break;
   }
}

Reply:
Made this change.